From 000ef7e7122f52f16b645d4c2e63fd7096448d9c Mon Sep 17 00:00:00 2001
From: Romain Francoise <rfrancoise@debian.org>
Date: Thu, 24 Feb 2011 23:22:49 +0100
Subject: [PATCH 02/10] `mail-from-style' is set to 'system-default by default

Signed-off-by: Julien Danjou <julien@danjou.info>
---
 lisp/mail/sendmail.el |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lisp/mail/sendmail.el b/lisp/mail/sendmail.el
index 331754f..15a177f 100644
--- a/lisp/mail/sendmail.el
+++ b/lisp/mail/sendmail.el
@@ -55,7 +55,7 @@
   :type 'file)
 
 ;;;###autoload
-(defcustom mail-from-style 'default
+(defcustom mail-from-style 'system-default
   "Specifies how \"From:\" fields look.
 
 If `nil', they contain just the return address like:
-- 
1.7.10.4

