From e5aadbd78e3342023ad237d0762b3764096aa992 Mon Sep 17 00:00:00 2001
From: Julien Danjou <julien@danjou.info>
Date: Fri, 25 Feb 2011 12:54:05 +0100
Subject: [PATCH 04/10] Debian-related items were added to the Help menu

---
 lisp/menu-bar.el |   33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/lisp/menu-bar.el b/lisp/menu-bar.el
index 88e59ef..ebf9a56 100644
--- a/lisp/menu-bar.el
+++ b/lisp/menu-bar.el
@@ -1623,6 +1623,28 @@ key, a click, or a menu-item"))
   (info "elisp")
   (Info-index topic))
 
+(defun debian-emacs-changelog ()
+  "Display the Debian changelog file for this package."
+  (interactive)
+  (view-file "/usr/share/doc/emacs-snapshot-common/changelog.Debian.gz")
+  (goto-address))
+
+(defun debian-emacs-news ()
+  "Display the Debian NEWS file for this package."
+  (interactive)
+  (view-file "/usr/share/doc/emacs-snapshot-common/NEWS.Debian.gz")
+  (goto-address))
+
+(defun debian-emacs-readme ()
+  "Display the Debian README file for this package."
+  (interactive)
+  ;; Force opening of this file in outline mode, even if the other mode
+  ;; is present.
+  (let ((auto-mode-alist (cons '("debian/README.Debian$" . outline-mode)
+			       auto-mode-alist)))
+    (view-file "/usr/share/doc/emacs-snapshot-common/README.Debian")
+    (goto-address)))
+
 (defvar menu-bar-search-documentation-menu
   (let ((menu (make-sparse-keymap "Search Documentation")))
 
@@ -1735,6 +1757,17 @@ key, a click, or a menu-item"))
       `(menu-item "Describe" ,menu-bar-describe-menu))
     (bindings--define-key menu [search-documentation]
       `(menu-item "Search Documentation" ,menu-bar-search-documentation-menu))
+    (bindings--define-key menu [sep2]
+      menu-bar-separator)
+    (bindings--define-key menu [debian-emacs-changelog]
+      `(menu-item ,(purecopy "Debian ChangeLog") debian-emacs-changelog
+                  :help ,(purecopy "Display the Debian changelog")))
+    (bindings--define-key menu [debian-emacs-news]
+      `(menu-item ,(purecopy "Debian News") debian-emacs-news
+                  :help ,(purecopy "Display the NEWS.Debian file")))
+    (bindings--define-key menu [debian-emacs-readme]
+      `(menu-item ,(purecopy "Debian README") debian-emacs-readme
+                  :help ,(purecopy "Display the README.Debian file")))
     (bindings--define-key menu [sep1]
       menu-bar-separator)
     (bindings--define-key menu [emacs-psychotherapist]
-- 
1.7.10.4

